File: example_ivol.py
"""
Demo benchmark for the inverse-Gaussian implied-volatility formula.

This is demo code for a Python-level comparison against py_vollib. It is not
the implementation used for the numerical results or speed benchmark reported
in the paper.

Paper:
An Explicit Solution to Black-Scholes Implied Volatility
Wolfgang Schadner
Available at: https://arxiv.org/abs/2604.24480

Copyright (c) 2026 Wolfgang Schadner.
All rights reserved.

No warranty is provided. This code is for research and demonstration purposes
only and should not be interpreted as production-ready numerical software.
"""

from __future__ import annotations

import argparse
import gc
from math import erfc
from pathlib import Path
from time import perf_counter

import numpy as np
from numba import njit
from py_vollib.black.implied_volatility import implied_volatility as lbr

DEFAULT_RUNS = 3
DEFAULT_REPS = 200
DEFAULT_OUTPUT = "iv_accuracy_speed_summary.txt"
P0 = np.nextafter(0.0, 1.0)
P1 = np.nextafter(1.0, 0.0)


@njit
def ncdf(x):
    return 0.5 * erfc(-x / 1.4142135623730951)


@njit
def ndtri(p):
    if p <= P0:
        p = P0
    if p >= P1:
        p = P1

    a0 = -3.969683028665376e01
    a1 = 2.209460984245205e02
    a2 = -2.759285104469687e02
    a3 = 1.383577518672690e02
    a4 = -3.066479806614716e01
    a5 = 2.506628277459239e00
    b0 = -5.447609879822406e01
    b1 = 1.615858368580409e02
    b2 = -1.556989798598866e02
    b3 = 6.680131188771972e01
    b4 = -1.328068155288572e01
    c0 = -7.784894002430293e-03
    c1 = -3.223964580411365e-01
    c2 = -2.400758277161838e00
    c3 = -2.549732539343734e00
    c4 = 4.374664141464968e00
    c5 = 2.938163982698783e00
    d0 = 7.784695709041462e-03
    d1 = 3.224671290700398e-01
    d2 = 2.445134137142996e00
    d3 = 3.754408661907416e00

    if p < 0.02425:
        q = np.sqrt(-2.0 * np.log(p))
        x = (((((c0 * q + c1) * q + c2) * q + c3) * q + c4) * q + c5) / (
            ((((d0 * q + d1) * q + d2) * q + d3) * q + 1.0)
        )
    elif p <= 0.97575:
        q = p - 0.5
        r = q * q
        x = (((((a0 * r + a1) * r + a2) * r + a3) * r + a4) * r + a5) * q / (
            (((((b0 * r + b1) * r + b2) * r + b3) * r + b4) * r + 1.0)
        )
    else:
        q = np.sqrt(-2.0 * np.log1p(-p))
        x = -(((((c0 * q + c1) * q + c2) * q + c3) * q + c4) * q + c5) / (
            ((((d0 * q + d1) * q + d2) * q + d3) * q + 1.0)
        )

    for _ in range(2):
        e = ncdf(x) - p
        u = e * 2.5066282746310002 * np.exp(0.5 * x * x)
        x -= u / (1.0 + 0.5 * x * u)

    return x


@njit
def bs_c(k, v):
    d1 = -k / v + 0.5 * v
    d2 = d1 - v
    return ncdf(d1) - np.exp(k) * ncdf(d2)


@njit
def fz(k, z, c):
    if z <= 0.0:
        return -c
    t = k / (2.8284271247461903 * z)
    return (
        ncdf(1.4142135623730951 * z - t)
        - np.exp(k) * ncdf(-1.4142135623730951 * z - t)
        - c
    )


@njit
def fpz(k, z):
    return 1.1283791670955126 * np.exp(0.5 * k - z * z - k * k / (16.0 * z * z))


@njit
def z_start(c):
    z = ndtri(0.5 * (1.0 + c)) / 1.4142135623730951
    if z < 1e-12:
        z = 1e-12
    return z


@njit
def z_root(k, c):
    if c <= P0:
        c = P0
    if c >= P1:
        c = P1

    z = z_start(c)
    f = fz(k, z, c)

    if f < 0.0:
        lo = z
        hi = 2.0 * z + 1e-12
        for _ in range(100):
            if fz(k, hi, c) >= 0.0:
                break
            lo = hi
            hi = 2.0 * hi + 1e-12
    else:
        hi = z
        lo = 0.5 * z
        for _ in range(100):
            if fz(k, lo, c) <= 0.0:
                break
            hi = lo
            lo = 0.5 * lo

    z = 0.5 * (lo + hi)

    for _ in range(30):
        f = fz(k, z, c)

        if f < 0.0:
            lo = z
        else:
            hi = z

        fp = fpz(k, z)
        zn = 0.5 * (lo + hi)

        if fp > 0.0 and np.isfinite(fp):
            r = f / fp
            h = -2.0 * z + k * k / (8.0 * z * z * z)
            den = 2.0 - r * h

            if den != 0.0 and np.isfinite(den):
                cand = z - 2.0 * r / den
                if cand > lo and cand < hi and np.isfinite(cand):
                    zn = cand

        z = zn

    return z


@njit
def iv_fig_numba(k, c):
    if abs(k) < 1e-12:
        return 2.0 * ndtri(0.5 * (1.0 + c))
    if k < 0.0:
        c = 1.0 + np.exp(-k) * (c - 1.0)
        k = -k
    return 2.8284271247461903 * z_root(k, c)


@njit
def iv_fig_batch(ks, cs):
    out = np.empty_like(ks)
    for i in range(ks.size):
        out[i] = iv_fig_numba(ks[i], cs[i])
    return out


def make_cases():
    vs = np.r_[0.01, 0.05 * np.arange(1, 41)]
    ds = np.array([0.05, 0.20, 0.30, 0.45, 0.55, 0.70, 0.80, 0.95], dtype=np.float64)
    grid_v, grid_d = np.meshgrid(vs, ds, indexing="ij")

    v = grid_v.ravel().astype(np.float64)
    d = grid_d.ravel().astype(np.float64)
    k = np.array([float(vv * (0.5 * vv - ndtri(float(dd)))) for vv, dd in zip(v, d)], dtype=np.float64)
    K = np.exp(k)
    c = np.array([float(bs_c(float(kk), float(vv))) for kk, vv in zip(k, v)], dtype=np.float64)

    return {
        "k": k,
        "K": K,
        "c": c,
        "v": v,
        "d": d,
    }


def iv_lbr_batch(cases):
    c = cases["c"]
    K = cases["K"]
    out = np.empty_like(c)
    for i in range(c.size):
        out[i] = float(lbr(float(c[i]), 1.0, float(K[i]), 0.0, 1.0, "c"))
    return out


def iv_fig_batch_py(cases):
    return iv_fig_batch(cases["k"], cases["c"])


def acc(name, vals, true):
    e = np.abs(vals - true)
    return {
        "method": name,
        "mean_abs_err": float(e.mean()),
        "max_abs_err": float(e.max()),
        "nan": int(np.isnan(vals).sum()),
    }


def speed(name, func, cases, runs, reps):
    func(cases)

    gc.disable()
    ts = []
    n = reps * cases["k"].size

    try:
        for _ in range(runs):
            t0 = perf_counter()
            checksum = 0.0
            for _ in range(reps):
                checksum += float(func(cases).sum())
            ts.append(perf_counter() - t0)
            if not np.isfinite(checksum):
                raise RuntimeError(f"{name} produced a non-finite checksum.")
    finally:
        gc.enable()

    ts = np.array(ts, dtype=np.float64)
    return {
        "method": name,
        "seconds_median": float(np.median(ts)),
        "seconds_mean": float(ts.mean()),
        "us_per_eval_median": float(1e6 * np.median(ts) / n),
        "us_per_eval_mean": float(1e6 * ts.mean() / n),
    }


def format_table(rows, columns):
    widths = {}
    for col in columns:
        values = [col] + [str(row[col]) for row in rows]
        widths[col] = max(len(v) for v in values)

    header = " ".join(col.ljust(widths[col]) for col in columns)
    body = [
        " ".join(str(row[col]).ljust(widths[col]) for col in columns)
        for row in rows
    ]
    return "\n".join([header] + body)


def summarize(cases, runs, reps, with_lbr):
    true = cases["v"]

    # Warm up the jitted path outside the timing loop.
    iv_fig_batch_py({key: value[:1] for key, value in cases.items()})

    accuracy_rows = []
    speed_rows = []

    fig_vals = iv_fig_batch_py(cases)
    accuracy_rows.append(acc("F_IG_numba_batch", fig_vals, true))
    speed_rows.append(speed("F_IG_numba_batch", iv_fig_batch_py, cases, runs, reps))

    if with_lbr:
        lbr_vals = iv_lbr_batch(cases)
        accuracy_rows.insert(0, acc("LBR_vollib", lbr_vals, true))
        speed_rows.insert(0, speed("LBR_vollib", iv_lbr_batch, cases, runs, reps))

    return accuracy_rows, speed_rows


def build_report(accuracy_rows, speed_rows):
    accuracy_columns = ["method", "mean_abs_err", "max_abs_err", "nan"]
    speed_columns = [
        "method",
        "seconds_median",
        "seconds_mean",
        "us_per_eval_median",
        "us_per_eval_mean",
    ]

    parts = [
        "Accuracy summary",
        format_table(accuracy_rows, accuracy_columns),
        "",
        "Speed summary",
        format_table(speed_rows, speed_columns),
    ]
    return "\n".join(parts)


def parse_args():
    parser = argparse.ArgumentParser(
        description="Demo benchmark for the inverse-Gaussian implied-volatility formula."
    )
    parser.add_argument("--runs", type=int, default=DEFAULT_RUNS, help="Number of timing runs.")
    parser.add_argument("--reps", type=int, default=DEFAULT_REPS, help="Repetitions per timing run.")
    parser.add_argument(
        "--output",
        type=Path,
        default=Path(DEFAULT_OUTPUT),
        help="Output path for the text summary.",
    )
    parser.add_argument(
        "--skip-lbr",
        action="store_true",
        help="Skip the slower py_vollib comparison and benchmark only the demo formula.",
    )
    return parser.parse_args()


def main():
    args = parse_args()
    cases = make_cases()
    accuracy_rows, speed_rows = summarize(
        cases=cases,
        runs=args.runs,
        reps=args.reps,
        with_lbr=not args.skip_lbr,
    )

    report = build_report(accuracy_rows, speed_rows)
    print(report)

    args.output.write_text(report + "\n", encoding="utf-8")
    print(f"\nSaved: {args.output}")


if __name__ == "__main__":
    main()

File: example_ivol.R
# Demo R implementation of the inverse-Gaussian implied-volatility formula.
#
# This is demonstration code for computing total Black-Scholes implied volatility
# from normalized call price and forward log-moneyness. It is not the optimized
# implementation used for the numerical speed results reported in the paper.
#
# Paper:
# An Explicit Solution to Black-Scholes Implied Volatility
# Wolfgang Schadner
# Available at: https://papers.ssrn.com/sol3/papers.cfm?abstract_id=6649499
#
# Copyright (c) 2026 Wolfgang Schadner.
# All rights reserved.
#
# This code is provided for research and demonstration purposes only, without
# warranty of any kind. It should not be interpreted as production-ready
# numerical software.

# Functions ---------------------------------------------------------------

iv_ig <- function(k, c, n = 30) {
  rt2 <- sqrt(2)
  rt8 <- sqrt(8)
  p0 <- .Machine$double.xmin
  p1 <- 1 - .Machine$double.eps
  twospi <- 2 / sqrt(pi)
  
  one <- function(k, c) {
    if (!is.finite(k) || !is.finite(c)) return(NaN)
    if (c <= 0) return(0)
    if (abs(k) < 1e-12) return(2 * qnorm(min(max(c + 0.5, p0), p1)))
    
    if (k < 0) {
      c <- 1 + exp(-k) * (c - 1)
      k <- -k
    }
    
    c <- min(max(c, p0), p1)
    
    fz <- function(z) {
      pnorm(rt2 * z - k / (2 * rt2 * z)) -
        exp(k) * pnorm(-rt2 * z - k / (2 * rt2 * z)) - c
    }
    
    z <- qnorm(min(max(0.5 * (1 + c), p0), p1)) / rt2
    if (!is.finite(z) || z < 1e-14) z <- 1e-14
    
    f <- fz(z)
    
    if (f < 0) {
      lo <- z
      hi <- 2 * z + 1e-14
      for (i in seq_len(100)) {
        if (fz(hi) >= 0) break
        lo <- hi
        hi <- 2 * hi + 1e-14
      }
    } else {
      hi <- z
      lo <- 0.5 * z
      for (i in seq_len(100)) {
        if (fz(lo) <= 0) break
        hi <- lo
        lo <- 0.5 * lo
      }
    }
    
    z <- 0.5 * (lo + hi)
    
    for (i in seq_len(n)) {
      f <- fz(z)
      if (f < 0) lo <- z else hi <- z
      
      fp <- twospi * exp(0.5 * k - z * z - k * k / (16 * z * z))
      zn <- 0.5 * (lo + hi)
      
      if (fp > 0 && is.finite(fp)) {
        r <- f / fp
        h <- -2 * z + k * k / (8 * z * z * z)
        den <- 2 - r * h
        if (den != 0 && is.finite(den)) {
          cand <- z - 2 * r / den
          if (cand > lo && cand < hi && is.finite(cand)) zn <- cand
        }
      }
      
      z <- zn
    }
    
    rt8 * z
  }
  
  m <- max(length(k), length(c))
  k <- rep(k, length.out = m)
  c <- rep(c, length.out = m)
  vapply(seq_len(m), function(i) one(k[i], c[i]), numeric(1))
}

bs_call <- function(k, v) {
  pnorm(-k / v + 0.5 * v) - exp(k) * pnorm(-k / v - 0.5 * v)
}


# Example -----------------------------------------------------------------

v_true <- c(0.01, seq(0.05, 2, 0.05))
d <- c(0.05, 0.20, 0.30, 0.45, 0.55, 0.70, 0.80, 0.95)

grid <- expand.grid(v = v_true, d = d)
grid$k <- with(grid, v * (0.5 * v - qnorm(d)))
grid$c <- with(grid, bs_call(k, v))
grid$v_hat <- with(grid, iv_ig(k, c))
grid$abs_err <- abs(grid$v_hat - grid$v)

summary <- data.frame(
  n = nrow(grid),
  mean_abs_err = mean(grid$abs_err),
  max_abs_err = max(grid$abs_err)
)

print(summary)
head(grid)
